wain-syntax-text/src/wat2wasm.rs

```diff
@@ -202,24 +206,57 @@ impl<'s> Context<'s> {
         }
     }
 
-    fn resolve_type_idx(&self, idx: wat::Index<'s>, offset: usize) -> Result<'s, u32> {
-        self.resolve_index(&self.type_indices, idx, offset, "type")
+    fn resolve_type_idx(&mut self, ty: &wat::TypeUse<'s>, offset: usize) -> Result<'s, u32> {
+        match &ty.idx {
+            wat::RefOrInline::Reference(idx) => {
+                let idx = self.resolve_index(&self.type_indices, idx, offset, "type")?;
+                if idx as usize >= self.types.len() {
+                    return Ok(idx);
+                }
+                let wat::FuncType {
+                    params, results, ..
+                } = &self.types[idx as usize];
+                if ty.params.is_empty() && ty.results.is_empty()
+                    || ty.params.len() == params.len()
+                        && ty.results.len() == results.len()
+                        && ty
+                            .params
+                            .iter()
+                            .zip(params.iter())
+                            .all(|(l, r)| l.ty == r.ty)
+                        && ty
+                            .results
+                            .iter()
+                            .zip(results.iter())
+                            .all(|(l, r)| l.ty == r.ty)
```

Comment:
```suggestion
if ty.params
        .iter()
        .map(|p| p.ty)
        .eq(params.iter().map(|p| p.ty))
        && ty
            .results
            .iter()
            .map(|r| r.ty)
            .eq(results.iter().map(|r| r.ty))
```

`.eq()` を使うとシーケンスの長さをチェックする必要がなくなるので，よりシンプルになりそうです．

Reply:
ちなみに、`is_empty()` は省略できないので残しました。

Reply:
あっ，すみません，そうですね．僕の suggestion のほうが間違いです